Resilience Metrics Reporting

xuanluo113 opened the pull request:
Pull request overview
---------------------
With increasing frequency and severity of extreme weather events (e.g., heat waves), it is crucial to ensure urban buildings and infrastructure are resilient to provide critical services to preserve human life and properties during natural disasters. Building resilience has the opportunity to become an additional value proposition for technologies and systems if it can be reliably quantified, valued, and trusted by stakeholders. Metrics or an assessment of potential vulnerability, likelihood, and consequence for each risk would help to prioritize further consideration of these risk factors of building resilience [1]. There is a need to measure the resilience of buildings to help owners make better decisions and protect their assets, to better assess the built environment resilience of larger geographic units such as communities and cities, and to complement existing assessments of building sustainability. Building performance simulation is a powerful tool to quantify building’s resilience and vulnerability under extreme weather events.

EnergyPlus currently calculates at each time-step a set of variables (e.g., zone/space air dry-bulb temperature, relative humidity, CO2 concentration, illuminance level) that can be further processed to generate reports on resilience metrics for thermal performance, visual performance, and indoor air quality (IAQ). For the spatial resolution, the reporting needs to consider both the entire building for aggregated results and individual spaces for detailed and worst-case results. For the temporal resolution, the reporting needs to consider the aggregated and average results as well as the worst-case results (e.g., maximum or minimum values at a specific time, worst day, worst week) in summer, winter, and the whole year. The resilience metrics target occupants’ health and wellbeing during extreme weather events. Therefore, the metrics involve the aggregations across space (occupants in the space) and time (e.g., X occupant·hours (Y%) with Heat Index between certain range for Zone Z). Post-processing to generate these reports requires EnergyPlus models to produce lots of time-step report variables, which can be time-consuming and can require significant storage. Hence, we propose to add a list of resilience-related tabular reports in EnergyPlus to provide a consistent and convenient path for users of resilience analysis through modeling and simulation.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

xuanluo113 committed:
NFP

xuanluo113 committed:
NFP

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into resilience-metric

xuanluo113 committed:
address NFP comments

hongtz68 committed:
Update Resilience Metrics Reporting.md

hongtz68 committed:
Update Resilience Metrics Reporting.md

xuanluo113 committed:
vars

xuanluo113 committed:
thermal

xuanluo113 committed:
Merge branch 'resilience-metric' of https://github.com/NREL/EnergyPlus into resilience-metric

xuanluo113 committed:
merge

xuanluo113 committed:
fix redundant calc

xuanluo113 committed:
co2 and visual

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into resilience-metric

xuanluo113 committed:
documentation

xuanluo113 committed:
fix test

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into resilience-metric

Myoldmopar commented on src/EnergyPlus/DataHeatBalFanSys.hh:
```diff
@@ -232,6 +232,24 @@ namespace DataHeatBalFanSys {
+    extern Array1D<Real64> ZoneHumidex;
+    extern Array1D_int ZoneNumOcc;
+
+    extern Array1D<std::vector<double>> ZoneHeatIndexHourBins;
```
Use `Real64` typedef not `double`

Myoldmopar commented on src/EnergyPlus/OutputReportTabular.hh:
```diff
@@ -213,6 +213,9 @@ namespace OutputReportTabular {
     extern bool displayEconomicResultSummary;
     extern bool displayHeatEmissionsSummary;
     extern bool displayEioSummary;
+    extern bool displayThermalResilienceSummary;
```
Are these reset in any clear_state calls?

xuanluo113 replied:
@Myoldmopar I suppose they are reset [here](https://github.com/NREL/EnergyPlus/blob/9805e969689297e1395bc106d9cc9ac4d1cbf0c5/src/EnergyPlus/OutputReportTabular.cc#L594).

Myoldmopar replied:
Got it, I zipped right past that.  OK, the problem is elsewhere.

Myoldmopar reviewed (changes requested):
I built this locally and verified the failing unit test suite immediately.  The unit test fails on an OutputReportTabular unit test, and I see you added some new global variables to that file, and I don't see those being reset in a clear_state.  I may have just missed the reset, in which case the problem is elsewhere, but it certainly seems suspicious.

Myoldmopar commented:
@xuanluo113 there is definitely something not being reset.  The integration test failure is failing on the TestRuntimeC test.  During this test, EnergyPlus is executed, reset, and then executed again.  When I run the unit test suite locally I see the same thing -- serially the unit test fails.  But when CI runs the unit tests, they don't fail, because it doesn't run them serially.  I'm building a debug build and should be able to find the problem quickly, but if not I'll just need you to debug in and find what isn't being reset here.

xuanluo113 commented:
> @xuanluo113 there is definitely something not being reset. The integration test failure is failing on the TestRuntimeC test. During this test, EnergyPlus is executed, reset, and then executed again. When I run the unit test suite locally I see the same thing -- serially the unit test fails. But when CI runs the unit tests, they don't fail, because it doesn't run them serially. I'm building a debug build and should be able to find the problem quickly, but if not I'll just need you to debug in and find what isn't being reset here.

@Myoldmopar Thanks so much. I didn't get the CI's complaints. My local builds, on both mac and ubuntu, do no reproduce the error so I kind of got stuck. Good to know and I can look at the issue in the reset direction as well.

Myoldmopar commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -4876,6 +4889,313 @@ namespace HeatBalanceSurfaceManager {
+        static std::vector<int> lowSETLongestStart(NumOfZones, 0.0);
+        static std::vector<int> highSETLongestStart(NumOfZones, 0.0);
+        static bool hasPierceSET = true;
+        static bool oneTimeFlag = true;
```
You **cannot** use static flags inside functions.  There's no way to reset this in a clear_state because it is hidden inside a function.  There are these two plus two inside other Report* functions you added here, and they have to be extracted out and become variables on the namespace, and then they need to be cleared in a clear_state call.  This should fix the problem.

xuanluo113 replied:
I see. That makes sense. Thanks!

xuanluo113 committed:
fix one time flag

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into resilience-metric

Myoldmopar commented:
@xuanluo113 this does look a lot better.  Locally running shows no problem.  Once CI starts working on your latest commit I anticipate it will be all green.  There are still a few static variables inside the functions, however.  Are you sure those don't need to be reset in a clear_state as well?

xuanluo113 commented:
@Myoldmopar I thought those were fine as those vectors are [reinitialized](https://github.com/NREL/EnergyPlus/blob/101849dd2a92326eb15a445e4c80a3a3a3dfc8f5/src/EnergyPlus/HeatBalanceSurfaceManager.cc#L4951) at the beginning of each simulation. But seems the API test still fails. Will try clearing those.

Myoldmopar commented:
No, CI hasn't run that latest commit yet.  Those results are for 9805e96.  However, there is now a conflict in the heat balance surface manager.  I suggest you wait until CI reports on 101849d, make sure it is happy, then move forward, addressing the conflict and any other identified issues.

xuanluo113 committed:
more docs and tests

Myoldmopar commented:
CI looks happy...just one report so far but still it's a good sign!

xuanluo113 committed:
merge conflict

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into resilience-metric

xuanluo113 commented:
Thanks. Just pushed another commit adding more unit tests and documentation, and merging conflicts.

Myoldmopar commented:
@xuanluo113 there are two issues going on.  One is that your new test file is failing with an array index out of bounds in a debug build.  Two is that this branch is now causing HospitalLowEnergy to time out again...we had that file running much faster, so I'm not sure if something in your branch is causing a huge slow-down or what is happening.

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into resilience-metric

xuanluo113 committed:
fix CI

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into resilience-metric

xuanluo113 commented:
@Myoldmopar Now CI turns green. The `HospitalLowEnergy` timeout was due to all the [warnings](https://github.com/NREL/EnergyPlus/blob/1426e755443ea4f8fe19497d4c7be554ac27a048/src/EnergyPlus/OutputReportTabular.cc#L11437) printed when `AllSummary` is declared in the model. When that's declared, the Thermal Resilience reports are required to be included in the tabular reports. But SET-hour report requires the People defined in a zone and Pierce thermal comfort method defined for the People. Similar to the Visual and CO2 reports. Actually all the "587 tests had: ERR diffs." is triggered by this, which does not look good. Apart from (1) do not report resilience tables when `AllSummary` is declared; (2) eliminate/reduce the warnings, do you have any better suggestions?

Myoldmopar commented:
@xuanluo113 so first, CI is clearly reporting to you that the TeX changes have error messages, and they are pointing you to the exact lines.  There are 3 instances of `&` in your changes that need to be escaped to be valid TeX.  

> The HospitalLowEnergy timeout was due to all the warnings printed when AllSummary is declared in the model.

I don't see how printing a warning could've caused the hospital file to time out, but maybe I'm misunderstanding.

> do you have any better suggestions?

It depends on the definition of `AllSummary`.  Does `AllSummary` mean that **every** single summary must be written?  Or does it mean that **every summary valid for the current simulation** must be written.  I'm inclined to say that you should just write the summaries that **are** valid, and ignore the ones that aren't.  Otherwise we need to modify every IDF to include those other objects, or we need to add a new key `AllSummaryExceptSomeStuff`.  I guess you could add a one time extra warning message that says "Some summaries could not be written because you didn't include necessary objects", but that's still going to add a warning to like every IDF output.

mjwitte commented:
@xuanluo113 @Myoldmopar I would suggest following the pattern used by the EconomicResultSummary report. It is activated by AllSummary* options, but it does not report if there are no tariffs, no warnings issued. See [code here.](https://github.com/NREL/EnergyPlus/blob/a6234498a7f4244983ac3d778025fa3e41b2c69a/src/EnergyPlus/EconomicTariff.cc#L4242-L4244) Would that approach work here? And this could be noted in the I/O ref output reports description for this report - "... is not produced if there are no ..... objects."

mjwitte commented:
Or another option is to produce the report without warnings and without content, but that seems silly. If this report is specifically requested, you could throw a warning if the required inputs are not present.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -4877,6 +4906,311 @@ namespace HeatBalanceSurfaceManager {
 
+    void ReportThermalResilience() {
+
+        Real64 c1 = -42.379;
+        Real64 c2 = 2.04901523;
+        Real64 c3 = 10.14333127;
+        Real64 c4 = -.22475541;
+        Real64 c5 = -.00683783;
+        Real64 c6 = -.05481717;
+        Real64 c7 = .00122874;
+        Real64 c8 = .00085282;
+        Real64 c9 = -.00000199;
```
What are these? Should have some comments explaining what these are and citing a source.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -4877,6 +4906,311 @@ namespace HeatBalanceSurfaceManager {
+                    if (People(iPeople).Pierce) {
+                        ZoneOccPierceSET(ZoneNum) = ThermalComfort::ThermalComfortData(iPeople).PierceSET;
+                    } else {
+                        ZoneOccPierceSET(ZoneNum) = NAN;
```
Do we really want to be setting NANs? This will halt in a debug config (at least the way the windows debug config is currently set up).

xuanluo113 replied:
@mjwitte This is to explicitly set the value to some kind of "null" instead of zero during runtime at certain time steps (no occupant). If that is not working, it would work the same to set it to some unreasonable values like "999" or "-999" for temperatures or add more logics with flags like `ZoneOccPierceSETIsNull` and `ZoneOccPierceSETLastTimeStepIsNull`.

Myoldmopar replied:
-999 is fine.  Or just -1 if that can be used as a sentinel?

mjwitte commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -11198,6 +11250,276 @@ namespace OutputReportTabular {
+
+            int columnNum = 5;
+            std::string subTitle = "Heat Index Hours";
+            std::vector<std::string> columnHeadNames = { "Safe (≤ 80°F) [Hours]",
+                                                         "Caution (80, 90°F] [Hours]",
+                                                         "Extreme Caution (90, 103°F] [Hours]",
+                                                         "Danger (103, 125°F] [Hours]",
+                                                         "Extreme Danger (> 125°F) [Hours]" };
+            WriteResilienceBinsTable(columnNum, tblTitle, subTitle, sqlTitle, columnHeadNames, ZoneHeatIndexHourBins);
```
@xuanluo113 This seems like the hard way to build a new table report instead of using the `OutputReportPredined` functions like `newPreDefReport` and `newPreDefColumn`.
Is this hard-wired to always write deg F and other data in deg C below? Output tables should be sensitive to units conversion options specified in OutputControl:Table:Style. @JasonGlazer ?

xuanluo113 replied:
I was following the style of [WriteAdaptiveComfortTable](https://github.com/NREL/EnergyPlus/blob/1426e755443ea4f8fe19497d4c7be554ac27a048/src/EnergyPlus/OutputReportTabular.cc#L11157) first and was not aware of the `OutputReportPredined` style. I could clean them up to follow the predefined style then.

There's nothing to do with deg F or C. The standard [equations](https://www.wpc.ncep.noaa.gov/html/heatindex_equation.shtml) and [thresholds](https://www.weather.gov/ama/heatindex) are in deg F, so the table headers are written this way. The calculation results are converted back to [deg C](https://github.com/NREL/EnergyPlus/blob/1426e755443ea4f8fe19497d4c7be554ac27a048/src/EnergyPlus/HeatBalanceSurfaceManager.cc#L4963) at the time step loop.

JasonGlazer replied:
The column headings having ranges of values in F is an unusual case. I would think the SI version of the table the values should be shown in ranges of C instead of F. You might need to be sensitive to the unit system requested by the user and change the values when defining the table.

xuanluo113 replied:
@JasonGlazer Thanks for the clarification. I'll convert the headers to deg C thresholds. 
Regarding the predefined table style, if I want to add Zone Min, Max, Average and Sum of a runtime variable to the table end, is it the best practice to calculate these aggregation values in the [FillRemainingPredefinedEntries](https://github.com/NREL/EnergyPlus/blob/1426e755443ea4f8fe19497d4c7be554ac27a048/src/EnergyPlus/OutputReportTabular.cc#L6447) function?

xuanluo113 commented:
@Myoldmopar @mjwitte Thanks for the clarification. The warnings were quite verbose (perhaps causing the timeout), printing each specific Zone with no People - Pierce model, and with no `Daylighting:Control` when visual resilience is declared.
I would go with Mike's second solution, as I feel it's necessary to give a warning when users specific require a report, and it's okay to be silent otherwise.

Another question I have is, I wanted to [set up the variable of Heat Index](https://github.com/NREL/EnergyPlus/blob/1426e755443ea4f8fe19497d4c7be554ac27a048/src/EnergyPlus/HeatBalanceSurfaceManager.cc#L4942) only when Thermal Resilience Report is declared. But [this flag](https://github.com/NREL/EnergyPlus/blob/1426e755443ea4f8fe19497d4c7be554ac27a048/src/EnergyPlus/OutputReportTabular.cc#L2075) is read in `UpdateTabularReports`, which is called 
> at the end of the time step loop and updates the arrays of data that will later be put into the tabular reports.

Seems if only weather file run period is defined, the Zone Heat Index is not showing in the RDD for some reason. Is there any better way to prefetch this flag? Thanks for helping out.

xuanluo113 committed:
fix-1

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into resilience-metric

xuanluo113 committed:
to predefined

xuanluo113 commented:
@Myoldmopar @mjwitte @JasonGlazer Seems the CI shows green for the latest commit. A few updates addressing the previous comments.

1. Fixed NaN.
2. Fixed the I/O documentation compiling error.
3. Display warnings only when resilience flags are explicitly declared.
4. Converted the tables to the predefined style.
5. Always calculate the Heat Index and Humidex and set the RDDs. (RDD diffs)

A remaining issue is some extra warnings regarding the report header units, like:
```
+   ** Warning ** Unable to find a unit conversion from Caution (26.7, 32.2°C] [Hours] into IP units
+   **   ~~~   ** Applying default conversion factor of 1.0
```
Is it for the brackets and parentheses used here? Seems [this](https://github.com/NREL/EnergyPlus/blob/2ef4fbe91abb3699c687a1920db0eef78619e379/src/EnergyPlus/OutputReportTabular.cc#L16630) function is trying the evaluate "Hours]" after parsing.

JasonGlazer commented on src/EnergyPlus/OutputReportPredefined.cc:
```diff
@@ -2650,6 +2768,75 @@ namespace OutputReportPredefined {
+        pdchHIHourCaution = newPreDefColumn(pdstHIHours, "Caution (26.7, 32.2°C] [Hours]");
+        pdchHIHourExtremeCaution = newPreDefColumn(pdstHIHours, "Extreme Caution (32.2, 39.4°C] [Hours]");
+        pdchHIHourDanger = newPreDefColumn(pdstHIHours, "Danger (39.4, 51.7°C] [Hours]");
+        pdchHIHourExtremeDanger = newPreDefColumn(pdstHIHours, "Extreme Danger (> 51.7°C) [Hours]");
```
The headers for these columns are only in SI units and won't be getting converted automatically. Maybe the newPreDefColumn needs to have an new optional third argument which is the IP version of the column header.

JasonGlazer commented on src/EnergyPlus/OutputReportPredefined.cc:
```diff
@@ -2650,6 +2768,75 @@ namespace OutputReportPredefined {
+        pdchHIOccuHourCaution = newPreDefColumn(pdstHIOccuHours, "Caution (26.7, 32.2°C] [OccupantHours]");
+        pdchHIOccuHourExtremeCaution = newPreDefColumn(pdstHIOccuHours, "Extreme Caution (32.2, 39.4°C] [OccupantHours]");
+        pdchHIOccuHourDanger = newPreDefColumn(pdstHIOccuHours, "Danger (39.4, 51.7°C] [OccupantHours]");
+        pdchHIOccuHourExtremeDanger = newPreDefColumn(pdstHIOccuHours, "Extreme Danger (> 51.7°C) [OccupantHours]");
```
What is difference between Hours and OccupantHours? OccupantHours is a new unit and we try to avoid adding them unless we have a good reason.

xuanluo113 replied:
Occupant * hours. So the unit should be [hrs] I guess. I'll change them to be consistent.

JasonGlazer commented on src/EnergyPlus/OutputReportPredefined.cc:
```diff
@@ -2650,6 +2768,75 @@ namespace OutputReportPredefined {
+
+        pdstHeatingSETHours = newPreDefSubTable(pdrThermalResilience, "Heating SET Hours");
+        pdchHeatingSETHours = newPreDefColumn(pdstHeatingSETHours, "SET ≤ 12.2°C Hours (°C)");
+        pdchHeatingSETOccuHours = newPreDefColumn(pdstHeatingSETHours, "SET ≤ 12.2°C OccupantHours (°C)");
```
I assume these are also hours, please add [hr]. Actually in general, all the "[Hours]" should just be "[hr]"

xuanluo113 replied:
Okay. Thanks.

Nigusse replied:
Resilience-Metrics table summary outputs format all look fine. Below is a ample for quick look.

![SampleResilienceTableSummary](https://user-images.githubusercontent.com/5805126/90674933-d302cf00-e227-11ea-8762-293ec49e777c.png)

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into resilience-metric

xuanluo113 committed:
unit fix

xuanluo113 committed:
unit

JasonGlazer commented:
@xuanluo113 as far as performance, you could change the calculations for the heat index and humidity index so they only occur when someone requests the variables or the tabular report that shows them.

xuanluo113 commented:
@JasonGlazer I wanted to [set up the variable of Heat Index](https://github.com/NREL/EnergyPlus/blob/1426e755443ea4f8fe19497d4c7be554ac27a048/src/EnergyPlus/HeatBalanceSurfaceManager.cc#L4942) only when Thermal Resilience Report is declared. But [this flag](https://github.com/NREL/EnergyPlus/blob/1426e755443ea4f8fe19497d4c7be554ac27a048/src/EnergyPlus/OutputReportTabular.cc#L2075) is read in `UpdateTabularReports`, which is called 
> at the end of the time step loop and updates the arrays of data that will later be put into the tabular reports.

Seems if only weather file run period is defined, the Zone Heat Index is not showing in the RDD for some reason. Do you know if there's a better way to prefetch this flag? Thanks for helping out.

JasonGlazer commented:
The output variable setup should always be done so that it shows up in the RDD but if the output variable is not requested and the table is not requested, you can skip the computation of the metric.

xuanluo113 commented:
@JasonGlazer That makes sense. Thank you!

xuanluo113 committed:
merge

xuanluo113 committed:
reduce heat index calc

xuanluo113 commented:
@JasonGlazer @Myoldmopar The remaining issue was fixed and the branch is up-to-date.

Nigusse commented on doc/engineering-reference/CMakeLists.txt:
```diff
@@ -148,6 +148,8 @@ SET( INCLUDED_TEX
   ${PROJECT_SOURCE_DIR}/engineering-reference/src/special-modules-reporting/zone-component-loads-summary.tex
   ${PROJECT_SOURCE_DIR}/engineering-reference/src/special-modules-reporting/output-table-annual.tex
+  ${PROJECT_SOURCE_DIR}/engineering-reference/src/special-modules-reporting/heat-emissions.tex
+  ${PROJECT_SOURCE_DIR}/engineering-reference/src/special-modules-reporting/resilience-metrics.tex
```
I see you have included *resilience-metrics* documentation but I don't see the *Resilience Metrics* section when I compile the engineering-reference project.  @xuanluo113 did you compile and confirmed that the *Resilience Metrics*  section is in the pdf?

Nigusse commented on doc/input-output-reference/src/overview/group-thermal-zone-description-geometry.tex:
```diff
@@ -4945,6 +4945,8 @@ \subsection{Zone Thermal Output(s)}\label{zone-thermal-outputs-1}
 Zone,Average,Zone Mean Radiant Temperature [C]
 Zone,Average,Zone Operative Temperature [C]
+Zone,Average,Zone Heat Index [C]
+Zone,Average,Zone Humidity Index [C]
```
OK. I see you have added these two zone level temporal variables.

Nigusse replied:
The RDD file shows these two zone level report variables:

```
Output:Variable,*,Zone Heat Index,hourly; !- Zone Average [C]
Output:Variable,*,Zone Humidity Index,hourly; !- Zone Average []
```

Nigusse commented on doc/input-output-reference/src/standard-output-reports/output-table-summaryreports.tex:
```diff
@@ -429,6 +429,17 @@ \subsubsection{Predefined Annual Summary Reports}\label{predefined-annual-summar
+
+\paragraph{Visual Resilience Summary}\label{VisualResilienceSummary}
+
+The Visual Resilience Summary provides the indoor illuminance level unmet hours and occupant$·$hours summary report for resilience. This includes the Illuminance Level Hours and OccupantHours reports. To activate the indoor illuminance calculation in EnergyPlus, users need to define the \textbf{Daylighting:Controls} and the \textbf{Daylighting:ReferencePoint} objects, even if no daylighting contr ...
```
Compiled the input-out-put reference documentation and confirmed that this new section is included.

Nigusse commented on src/EnergyPlus/DataHeatBalFanSys.cc:
```diff
@@ -240,6 +240,23 @@ namespace DataHeatBalFanSys {
+    Array1D<Real64> ZoneOccPierceSETLastStep;
+    Array1D<std::vector<Real64>> ZoneLowSETHours; // LowSETHour, LowSETOccHour, LowSetLongestDur, LowSetLongestTimeStamp
+    Array1D<std::vector<Real64>> ZoneHighSETHours; //HighSETHour, HighSETOccHour, HighSetLongestDur, HighSetLongestTimeStamp
+
```
All real vectors are now defined as *Real64* as suggested in earlier review comment.

Nigusse commented on src/EnergyPlus/OutputReportPredefined.cc:
```diff
@@ -2650,6 +2768,75 @@ namespace OutputReportPredefined {
+        pdchHumidexOccuHourSome = newPreDefColumn(pdstHumidexOccuHours, "Some Discomfort (> 29, ≤ 40) [hr]");
+        pdchHumidexOccuHourGreat = newPreDefColumn(pdstHumidexOccuHours, "Great Discomfort; Avoid Exertion (> 40, ≤ 45) [hr]");
+        pdchHumidexOccuHourDanger = newPreDefColumn(pdstHumidexOccuHours, "Dangerous (> 45, ≤ 50) [hr]");
+        pdchHumidexOccuHourStroke = newPreDefColumn(pdstHumidexOccuHours, "Heat Stroke Quite Possible (> 50) [hr]");
```
Now the heading now replaced with [hr] instead of [hours] per previous review comments. And temperatures in summary tables heading are in Celsius scale.

Nigusse commented on src/EnergyPlus/OutputReportPredefined.cc:
```diff
@@ -2650,6 +2768,75 @@ namespace OutputReportPredefined {
+        pdchIllumOccuHourDim = newPreDefColumn(pdstIllumOccuHours, "Dim (> 100, <= 300 lux) [hr]");
+        pdchIllumOccuHourAdequate = newPreDefColumn(pdstIllumOccuHours, "Adequate (> 300, <= 500 lux) [hr]");
+        pdchIllumOccuHourBright = newPreDefColumn(pdstIllumOccuHours, "Bright (>500 lux) [hr]");
+
```
Resilience-Metrics IAQ and Illuminance table summary outputs format look fine. Below is a ample for quick look.

![SampleResilienceIAQIlluminanceSummary](https://user-images.githubusercontent.com/5805126/90675876-114cbe00-e229-11ea-9941-1d53d89ed76c.png)

Commit:
Merge develop; resolved conflicts

Nigusse commented on tst/EnergyPlus/unit/HeatBalanceSurfaceManager.unit.cc:
```diff
@@ -2936,7 +3087,5 @@ TEST_F(EnergyPlusFixture, HeatBalanceSurfaceManager_TestInitHBInterzoneWindow)
     EXPECT_EQ(1, DataHeatBalSurface::FractDifShortZtoZ(1, 1));
-
 }
-
```
Merged in develop and resolved the conflict.

Nigusse replied:
Built and ran the branch with develop pulled in and all unit tests passed.

Nigusse commented on src/EnergyPlus/OutputReportTabular.hh:
```diff
@@ -820,6 +825,20 @@ namespace OutputReportTabular {
+    void WriteSETHoursTable(int const columnNum,
+                            std::vector<std::string> const &columnHead,
+                            Array1D<std::vector<double>> const &ZoneBins);
+
```
@xuanluo113 Please replace double with Real64 per previous review comment from @Myoldmopar.

Nigusse reviewed (commented):
The code change and unit test look good. Pulled in develop, built and ran the branch.  All unit test passed. I have a couple comments that @xuanluo113 needs to address: 
-  Complied the Eng/Ref documentation but I could not see/find the Resilience Metric Sub-section in the pdf file.
- a couple of variable declaration still use *double* instead of *Real64*. See review comments. 

The updated branch got pushed. In the meantime I will test the new example file.

Nigusse commented:
The following three tabular summary reports will not be generated unless specified in *Output:Table:SummaryReports* object:

```
- Annual Thermal Resilience Summary
- Annual CO2 Resilience Summary
- Annual Visual Resilience Summary
```

Confirmed by commenting out and running the in the example file provided. 
```
  Output:Table:SummaryReports,
    AnnualBuildingUtilityPerformanceSummary,  !- Report 1 Name
    DemandEndUseComponentsSummary;            !- Report 2 Name
!    ThermalResilienceSummary,                 !- Report 3 Name
!    CO2ResilienceSummary,                     !- Report 4 Name
!    VisualResilienceSummary;                  !- Report 5 Name

```

Myoldmopar commented:
It's disappointing that the `double` declarations weren't cleaned up.  @Nigusse thanks for taking a look at this.  It looks like it is good outside of the `double` declarations, and the documentation isn't showing up because the new TeX file was never included in the table of contents.  It should've been listed [here](https://github.com/NREL/EnergyPlus/blob/6806643a4f2a1bd773f1aab83ba3d3e2306cfc3d/doc/engineering-reference/engineering-reference.tex#L306).

If this continues to come back clean then I won't waste more CI cycles right now.  Instead I'll open an issue that @xuanluo113 will need to address by cleaning up the double declarations and ensuring the new documentation shows up in the engineering reference.

Myoldmopar commented:
Built and tested locally with develop; all is well.  CI diffs are as expected.  I'm going to merge this and open up a follow-up issue.
